File: sdk/typescript/src/http.ts
// The transport: the one place that knows about auth, retries, timeouts and error mapping.
//
// The 38 generated operation methods do nothing but describe a request and hand it here, so any
// change to HOW the SDK talks to a deployment is a change to this file alone.

import {
  CatFactoryConnectionError,
  CatFactoryDecodeError,
  CatFactoryError,
  CatFactoryTimeoutError,
  toApiError,
} from './errors.ts'
import { type EventStream, readEventStream } from './sse.ts'

/** Per-call overrides. Every operation method takes one as its last argument. */
export interface RequestOptions {
  /** Abort the call from the outside (composed with the client's own timeout). */
  signal?: AbortSignal
  /** Override the client's `timeoutMs` for this call. `0` disables the deadline. */
  timeoutMs?: number
  /** Extra headers, merged over the client's. */
  headers?: Record<string, string>
  /**
   * Override the retry budget for this call. Note the default policy below: a non-idempotent
   * request is never retried automatically, so raising this does not make `POST /jobs`
   * retry-safe.
   */
  maxRetries?: number
}

export interface ClientOptions {
  /** The deployment's origin, e.g. `https://cat-factory.example.com`. */
  baseUrl: string
  /** A public-API key: `cf_live_<keyId>.<secret>`. */
  apiKey: string
  /** Per-request deadline in ms; `0` disables it. Default 30_000. */
  timeoutMs?: number
  /** Retries for a RETRIABLE failure (see `isRetriable`). Default 2. */
  maxRetries?: number
  /** Headers sent on every request. */
  headers?: Record<string, string>
  /** Swap the HTTP implementation (a proxy agent, a test double). Defaults to global `fetch`. */
  fetch?: typeof globalThis.fetch
  /** Prefixed to `User-Agent`, so a deployment's logs can attribute calls to your integration. */
  userAgent?: string
}

export interface RequestSpec {
  method: string
  path: string
  body?: unknown
  query?: Record<string, unknown>
  options: RequestOptions
}

/** SDK version, stamped into `User-Agent`. Kept in step with package.json by `check:sdk`. */
export const SDK_VERSION = '0.6.0'

/**
 * Percent-encode a path parameter.
 *
 * `encodeURIComponent` and not a raw interpolation: an id is server-supplied but travels through
 * a caller's own storage, and one carrying a `/` or a `?` would otherwise silently re-target the
 * request at a different route rather than 404 on the id it names.
 */
export function encodePathSegment(value: string): string {
  return encodeURIComponent(value)
}

/** Whether a failed attempt may be replayed. */
function isRetriable(method: string, status: number | null): boolean {
  // A transport failure with no response (status null) tells us nothing about whether the
  // server acted, so only a method that is idempotent BY DEFINITION may be replayed. `POST
  // /jobs` and `POST /tasks/:id/start` both cost real LLM work, and a duplicate is not
  // something the SDK may decide to risk on the caller's behalf.
  const idempotent = method === 'GET' || method === 'HEAD' || method === 'DELETE'
  if (!idempotent) return false
  if (status === null) return true
  return status === 429 || status === 502 || status === 503 || status === 504
}

/** Full jitter on an exponential base, so a fleet of clients does not retry in lockstep. */
function backoffMs(attempt: number): number {
  return Math.round(Math.random() * Math.min(8_000, 250 * 2 ** attempt))
}

const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms))

/** Serialize the query bag, dropping absent values so `?limit=undefined` is impossible. */
function buildQuery(query: Record<string, unknown> | undefined): string {
  if (!query) return ''
  const params = new URLSearchParams()
  for (const [key, value] of Object.entries(query)) {
    if (value === undefined || value === null) continue
    params.append(key, String(value))
  }
  const rendered = params.toString()
  return rendered ? `?${rendered}` : ''
}

export class Transport {
  private readonly baseUrl: string
  private readonly apiKey: string
  private readonly timeoutMs: number
  private readonly maxRetries: number
  private readonly headers: Record<string, string>
  private readonly doFetch: typeof globalThis.fetch

  constructor(options: ClientOptions) {
    if (!options.baseUrl) throw new Error('cat-factory SDK: `baseUrl` is required.')
    if (!options.apiKey) throw new Error('cat-factory SDK: `apiKey` is required.')
    this.baseUrl = options.baseUrl.replace(/\/+$/, '')
    this.apiKey = options.apiKey
    this.timeoutMs = options.timeoutMs ?? 30_000
    this.maxRetries = options.maxRetries ?? 2
    const agent = options.userAgent ? `${options.userAgent} ` : ''
    this.headers = {
      accept: 'application/json',
      'user-agent': `${agent}cat-factory-sdk-js/${SDK_VERSION}`,
      ...options.headers,
    }
    this.doFetch = options.fetch ?? globalThis.fetch.bind(globalThis)
  }

  /** Perform a request, returning the decoded JSON body. */
  async request<T>(spec: RequestSpec): Promise<T> {
    const response = await this.send(spec, 'application/json')
    const text = await response.text()
    if (text.length === 0) return undefined as T
    try {
      return JSON.parse(text) as T
    } catch (cause) {
      throw new CatFactoryDecodeError(
        `cat-factory SDK: ${spec.method} ${spec.path} returned a body that is not JSON.`,
        text,
        { cause },
      )
    }
  }

  /** Perform a request whose success carries no body (a 204). */
  async requestNoContent(spec: RequestSpec): Promise<void> {
    const response = await this.send(spec, 'application/json')
    // Drain, so a keep-alive connection is returned to the pool rather than held open.
    await response.arrayBuffer()
  }

  /**
   * Open a server-sent event stream. Deliberately NOT retried: a reconnect would replay the
   * stream from its start, and the caller — who knows which events it has already acted on —
   * is the only party that can decide whether that is safe.
   */
  async stream(spec: RequestSpec): Promise<EventStream> {
    const response = await this.send(
      { ...spec, options: { ...spec.options, maxRetries: 0 } },
      'text/event-stream',
    )
    if (!response.body) {
      throw new CatFactoryConnectionError('cat-factory SDK: the event stream carried no body.')
    }
    return readEventStream(response.body)
  }

  private async send(spec: RequestSpec, accept: string): Promise<Response> {
    const url = `${this.baseUrl}${spec.path}${buildQuery(spec.query)}`
    const budget = spec.options.maxRetries ?? this.maxRetries
    let lastError: unknown

    for (let attempt = 0; ; attempt += 1) {
      const timeoutMs = spec.options.timeoutMs ?? this.timeoutMs
      const controller = new AbortController()
      const onAbort = (): void => controller.abort(spec.options.signal?.reason)
      spec.options.signal?.addEventListener('abort', onAbort, { once: true })
      const timer =
        timeoutMs > 0 ? setTimeout(() => controller.abort(new DeadlineReached()), timeoutMs) : null

      try {
        const response = await this.doFetch(url, {
          method: spec.method,
          headers: {
            // Client headers, then per-call ones, then the three the SDK owns — which therefore
            // win. An `authorization` the transport did not build, or an `accept` that disagrees
            // with how the response is about to be read, are not customisations; they are the
            // client not working. All four SDKs apply this same precedence.
            ...this.headers,
            ...spec.options.headers,
            accept,
            authorization: `Bearer ${this.apiKey}`,
            ...(spec.body === undefined ? {} : { 'content-type': 'application/json' }),
          },
          body: spec.body === undefined ? undefined : JSON.stringify(spec.body),
          signal: controller.signal,
        })
        if (response.ok) return response

        const requestId = response.headers.get('x-request-id')
        if (attempt < budget && isRetriable(spec.method, response.status)) {
          // Honour `Retry-After` when the server states one: it is the deployment's own
          // knowledge of when the limit clears, which beats our blind backoff curve.
          await sleep(retryAfterMs(response) ?? backoffMs(attempt))
          continue
        }
        throw toApiError(response.status, await readBodySafely(response), requestId)
      } catch (error) {
        // The CALLER's cancellation is checked FIRST, and on the signal rather than on the shape
        // of the error: `abort(reason)` rejects the fetch with that reason verbatim, so a caller
        // who aborts with a plain `new Error('user navigated away')` produces something whose
        // `name` is not `AbortError`. Gating on the name alone let exactly that case fall through
        // to the retry branch below — a cancelled GET was replayed to the budget and then
        // reported as a connection failure, which is neither what happened nor what was asked
        // for. A cancellation is the outcome the caller chose; it is never retried, and never
        // re-wrapped.
        if (spec.options.signal?.aborted) throw spec.options.signal.reason ?? error
        if (error instanceof Error && error.name === 'AbortError') {
          // Ours, then: the deadline. Distinct from the above because a timeout is something the
          // caller may want to retry with a longer budget.
          throw new CatFactoryTimeoutError(
            `cat-factory SDK: ${spec.method} ${spec.path} exceeded ${timeoutMs}ms.`,
            { cause: error },
          )
        }
        // An error we already classified (an API refusal) propagates untouched.
        if (isSdkError(error)) throw error
        lastError = error
        if (attempt < budget && isRetriable(spec.method, null)) {
          await sleep(backoffMs(attempt))
          continue
        }
        throw new CatFactoryConnectionError(
          `cat-factory SDK: ${spec.method} ${spec.path} failed to reach ${this.baseUrl}.`,
          { cause: lastError },
        )
      } finally {
        if (timer) clearTimeout(timer)
        spec.options.signal?.removeEventListener('abort', onAbort)
      }
    }
  }
}

/** Marker carried by the abort reason our own deadline raises. */
class DeadlineReached extends Error {
  override readonly name = 'AbortError'
}

/**
 * An error this SDK already classified. It propagates untouched rather than being re-wrapped as
 * a connection failure: an API refusal that reached us through the `catch` below is a verdict,
 * not a transport fault, and re-wrapping it would hide the status the caller needs.
 */
function isSdkError(error: unknown): boolean {
  return error instanceof CatFactoryError
}

/** Read a failed response's body without letting a decode fault mask the real failure. */
async function readBodySafely(response: Response): Promise<unknown> {
  const text = await response.text().catch(() => '')
  if (!text) return null
  try {
    return JSON.parse(text)
  } catch {
    return text
  }
}

/** `Retry-After` in ms — seconds or an HTTP date — or null when absent/unparsable. */
function retryAfterMs(response: Response): number | null {
  const header = response.headers.get('retry-after')
  if (!header) return null
  const seconds = Number(header)
  if (Number.isFinite(seconds) && seconds >= 0) return Math.min(seconds * 1000, 60_000)
  const date = Date.parse(header)
  if (Number.isNaN(date)) return null
  return Math.max(0, Math.min(date - Date.now(), 60_000))
}

File: sdk/python/cat_factory/_http.py
"""The transport: the one place that knows about auth, retries, timeouts and error mapping.

Built on ``urllib`` from the standard library rather than ``requests`` or ``httpx``. A client
library's dependencies become every consumer's dependencies, and a transitive HTTP stack is the
single most common source of version conflicts in a Python application --- so this SDK installs
with none. A caller who wants connection pooling or HTTP/2 supplies their own opener through
``transport_factory`` rather than being forced into ours.
"""

from __future__ import annotations

import datetime
import email.utils
import json
import random
import time
import urllib.error
import urllib.request
from collections.abc import Callable, Mapping
from typing import Any
from urllib.parse import urlencode

from ._sse import EventStream
from .errors import (
    CatFactoryConnectionError,
    CatFactoryDecodeError,
    CatFactoryError,
    CatFactoryTimeoutError,
    to_api_error,
)

#: SDK version, stamped into ``User-Agent``. Kept in step with pyproject.toml by ``check:sdk``.
SDK_VERSION = "0.4.0"

#: Methods that may be replayed after a failure.
#:
#: A transport failure with no response tells us nothing about whether the server acted, so only
#: a method that is idempotent BY DEFINITION is replayed. ``POST /jobs`` and
#: ``POST /tasks/:id/start`` both cost real LLM work, and a duplicate is not something the SDK
#: may decide to risk on the caller's behalf.
_IDEMPOTENT = frozenset({"GET", "HEAD", "DELETE"})

#: Statuses worth a second attempt when the method allows it.
_RETRIABLE_STATUS = frozenset({429, 502, 503, 504})


def _backoff_seconds(attempt: int) -> float:
    """Full jitter on an exponential base, so a fleet of clients does not retry in lockstep."""
    return random.random() * min(8.0, 0.25 * (2**attempt))


class Transport:
    """Performs requests against a deployment, applying auth, deadlines and the retry policy."""

    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        timeout: float = 30.0,
        max_retries: int = 2,
        headers: Mapping[str, str] | None = None,
        user_agent: str | None = None,
        opener: urllib.request.OpenerDirector | None = None,
    ) -> None:
        if not base_url:
            raise ValueError("cat-factory SDK: base_url is required.")
        if not api_key:
            raise ValueError("cat-factory SDK: api_key is required.")
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._timeout = timeout
        self._max_retries = max_retries
        agent = f"{user_agent} " if user_agent else ""
        self._headers = {
            "accept": "application/json",
            "user-agent": f"{agent}cat-factory-sdk-python/{SDK_VERSION}",
            **dict(headers or {}),
        }
        self._opener = opener or urllib.request.build_opener()

    # -- public surface the generated operations call ------------------------------------------

    def request(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        query: Mapping[str, Any] | None = None,
        timeout: float | None = None,
    ) -> Any:
        """Perform a request and return the decoded JSON body."""
        raw = self._send(method, path, body, query, timeout, "application/json")
        text = raw.decode("utf-8")
        if not text:
            return None
        try:
            return json.loads(text)
        except json.JSONDecodeError as exc:
            raise CatFactoryDecodeError(
                f"cat-factory SDK: {method} {path} returned a body that is not JSON.", text
            ) from exc

    def request_no_content(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        query: Mapping[str, Any] | None = None,
        timeout: float | None = None,
    ) -> None:
        """Perform a request whose success carries no body (a 204)."""
        self._send(method, path, body, query, timeout, "application/json")

    def request_bytes(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        query: Mapping[str, Any] | None = None,
        timeout: float | None = None,
    ) -> bytes:
        """Perform a request whose success carries BYTES rather than JSON.

        Returned whole rather than as a file object: the listing endpoint that hands out these
        ids also carries each artifact's exact ``byteSize``, so a caller decides whether to fetch
        BEFORE issuing the request, and every artifact is bounded by the platform's own upload
        ceiling.

        ``*/*``, because the endpoint declares SEVERAL media types (the image allow-list plus an
        octet-stream fallback) and answers with whichever one the stored artifact is; naming any
        single one would disagree with most of what it sends.
        """
        return self._send(method, path, body, query, timeout, "*/*")

    def stream(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        query: Mapping[str, Any] | None = None,
        timeout: float | None = None,
    ) -> EventStream:
        """Open a server-sent event stream.

        Deliberately NOT retried: a reconnect would replay the stream from its start, and the
        caller --- who knows which events it has already acted on --- is the only party that can
        decide whether that is safe.
        """
        request = self._build(method, path, body, query, "text/event-stream")
        try:
            # No `with`: the response IS the stream, and closing it here would close the socket
            # the caller is about to read from. `EventStream.close()` owns it from now on.
            raw = self._opener.open(request, timeout=timeout or self._timeout)
        except urllib.error.HTTPError as exc:
            raise self._from_http_error(exc) from exc
        except (urllib.error.URLError, TimeoutError, OSError) as exc:
            raise self._from_transport_error(exc, method, path) from exc
        _release_read_deadline(raw)
        return EventStream(raw)

    # -- internals -----------------------------------------------------------------------------

    def _url(self, path: str, query: Mapping[str, Any] | None) -> str:
        # Drop absent values so `?limit=None` is impossible.
        pairs = [(k, v) for k, v in (query or {}).items() if v is not None]
        rendered = urlencode([(k, _wire(v)) for k, v in pairs])
        return f"{self._base_url}{path}{'?' + rendered if rendered else ''}"

    def _build(
        self,
        method: str,
        path: str,
        body: Any,
        query: Mapping[str, Any] | None,
        accept: str,
    ) -> urllib.request.Request:
        payload = None if body is None else json.dumps(body).encode("utf-8")
        request = urllib.request.Request(self._url(path, query), data=payload, method=method)
        for key, value in self._headers.items():
            request.add_header(key, value)
        request.add_header("accept", accept)
        request.add_header("authorization", f"Bearer {self._api_key}")
        if payload is not None:
            request.add_header("content-type", "application/json")
        return request

    def _send(
        self,
        method: str,
        path: str,
        body: Any,
        query: Mapping[str, Any] | None,
        timeout: float | None,
        accept: str,
    ) -> bytes:
        budget = self._max_retries
        deadline = timeout if timeout is not None else self._timeout

        for attempt in range(budget + 1):
            request = self._build(method, path, body, query, accept)
            try:
                with self._opener.open(request, timeout=deadline) as response:
                    return response.read()
            except urllib.error.HTTPError as exc:
                retriable = method in _IDEMPOTENT and exc.code in _RETRIABLE_STATUS
                if attempt < budget and retriable:
                    # Honour `Retry-After` when the server states one: it is the deployment's own
                    # knowledge of when the limit clears, which beats a blind backoff curve.
                    time.sleep(_retry_after(exc) or _backoff_seconds(attempt))
                    continue
                raise self._from_http_error(exc) from exc
            except (urllib.error.URLError, TimeoutError, OSError) as exc:
                if attempt < budget and method in _IDEMPOTENT:
                    time.sleep(_backoff_seconds(attempt))
                    continue
                raise self._from_transport_error(exc, method, path) from exc
        raise AssertionError("unreachable: the retry loop always returns or raises")

    def _from_http_error(self, exc: urllib.error.HTTPError) -> CatFactoryError:
        raw = exc.read()
        try:
            parsed: Any = json.loads(raw.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError):
            parsed = raw.decode("utf-8", errors="replace")
        return to_api_error(exc.code, parsed, exc.headers.get("x-request-id"))

    def _from_transport_error(self, exc: Exception, method: str, path: str) -> CatFactoryError:
        # A timeout and a connection failure need DIFFERENT reactions --- one may succeed with a
        # longer budget, the other will not --- so they stay apart rather than collapsing into
        # one "request failed".
        if isinstance(exc, TimeoutError) or isinstance(getattr(exc, "reason", None), TimeoutError):
            return CatFactoryTimeoutError(
                f"cat-factory SDK: {method} {path} exceeded its deadline."
            )
        return CatFactoryConnectionError(
            f"cat-factory SDK: {method} {path} failed to reach {self._base_url} ({exc})."
        )


def _release_read_deadline(raw: Any) -> None:
    """Stop the client deadline once a STREAM's response headers are in.

    ``urlopen(timeout=...)`` sets a SOCKET timeout, which keeps applying to every subsequent read
    --- so on a stream it silently becomes an inactivity limit rather than a request deadline. The
    deployment writes an SSE frame only when the run's projection CHANGES and sends no heartbeat,
    and a parked run waits for a human indefinitely by design, so a quiet stream is the normal
    state of a healthy one: left in place, the deadline aborts exactly the runs a caller most
    wants to watch.

    So the deadline covers the wait for HEADERS and nothing after it, which is what the other
    three SDKs already mean by it (TypeScript clears its abort timer once fetch resolves, Go stops
    the stream's timer at the same point, and the JDK's ``HttpRequest.timeout`` is defined that
    way). A stream is instead bounded by the caller closing it --- and, above it, by the
    deployment's own connection cap.

    urllib exposes no public way to do this, so the socket is reached through the attribute chain
    CPython actually builds (``HTTPResponse.fp`` -> ``BufferedReader.raw`` -> ``SocketIO._sock``).
    A response that does not have that shape --- a caller-supplied opener, a future CPython ---
    keeps its socket timeout: batching or an inactivity bound is a far smaller harm than reaching
    into an object we did not recognise.
    """
    socket = getattr(getattr(getattr(raw, "fp", None), "raw", None), "_sock", None)
    settimeout = getattr(socket, "settimeout", None)
    if settimeout is not None:
        settimeout(None)


def _wire(value: Any) -> str:
    """Render a query value the way the server reads it (``true``, not Python's ``True``)."""
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(getattr(value, "value", value))


def _retry_after(exc: urllib.error.HTTPError) -> float | None:
    """``Retry-After`` in seconds, capped, or None when absent or unparsable.

    Both wire forms, because RFC 9110 allows either and a deployment behind a proxy or CDN
    routinely gets the HTTP-date one written for it. Reading only the numeric form silently
    discarded the deployment's own knowledge of when the limit clears and fell back to a blind
    backoff curve --- which still works, just worse, so nothing would ever have surfaced it.
    """
    header = exc.headers.get("retry-after")
    if not header:
        return None
    try:
        return max(0.0, min(float(header), 60.0))
    except ValueError:
        pass
    parsed = email.utils.parsedate_to_datetime(header.strip())
    if parsed is None:
        return None
    if parsed.tzinfo is None:
        # An HTTP-date is GMT by definition; a naive one has simply lost the marker.
        parsed = parsed.replace(tzinfo=datetime.timezone.utc)
    delta = (parsed - datetime.datetime.now(datetime.timezone.utc)).total_seconds()
    return max(0.0, min(delta, 60.0))


#: The type of a caller-supplied opener factory, for a consumer that wants their own HTTP stack.
TransportFactory = Callable[..., Transport]
